### PR TITLE
[lldb] Implement ShouldSkipValidation

### PR DESCRIPTION
**Pull request by augusto2112**
Unify check for tricky types to validate in one function.

### CODE REVIEW

File: lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp [hunk 1]
```diff
@@ -2653,7 +2643,8 @@ CompilerType TypeSystemSwiftTypeRef::GetChildCompilerTypeAtIndex(
   if (get_ast_num_children() <
       runtime->GetNumChildren({this, type}, valobj).getValueOr(0))
     return impl();
-
+  if (ShouldSkipValidation(type))
+    return impl();
```
Comment (adrian-prantl) [inline lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp:2640]: Why is this necessary? Isn't this check part of the VALIDATE_AND_RETURN macro below?

File: lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.h [hunk 2]
```diff
@@ -335,6 +335,10 @@ class TypeSystemSwiftTypeRef : public TypeSystemSwift {
   clang::api_notes::APINotesManager *
   GetAPINotesManager(ClangExternalASTSourceCallbacks *source, unsigned id);
 
+  /// Check whether the type being dealt with is tricky to validate due to
```
Comment (adrian-prantl) [inline lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.h:340]: Could we wrap the entire function in #ifndef NDEBUG so we don't accidentally call it in non-asserts builds?

File: lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp (hunk 1 above)
Comment (augusto2112) [reply lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp:2640]: There are a bunch of equivalence asserts right below this and before the VALIDATE_AND_RETURN macro.

File: lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.h (hunk 2 above)
Comment (augusto2112) [reply lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.h:340]: Done!

**Review by adrian-prantl — APPROVED** [review]

File: lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp [hunk 3]
```diff
@@ -3531,3 +3470,40 @@ TypeSystemSwiftTypeRef::GetGenericArgumentType(opaque_compiler_type_t type,
...
+   auto mangled_name = GetMangledTypeName(type);
+   if (mangled_name == "$sSo18NSNotificationNameaD" ||
+       mangled_name == "$s10Foundation12NotificationV4NameaD" ||
+       mangled_name == "$sSo9NSDecimalaD")
+     return true;
+
+  auto mangled_name = GetMangledTypeName(type);
+  if (mangled_name == "$sSo18NSNotificationNameaD" ||
+      mangled_name == "$sSo9NSDecimalaD")
+    return true;
```
Comment (kastiglione) [inline lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp]: this can be deleted

Comment (augusto2112) [reply lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp]: Whoops, I think this was as rebase that went wrong!

File: lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp [hunk 4]
```diff
@@ -3531,3 +3470,35 @@ TypeSystemSwiftTypeRef::GetGenericArgumentType(opaque_compiler_type_t type,
   VALIDATE_AND_RETURN(impl, GetGenericArgumentType, type, (ReconstructType(type), idx),
     (ReconstructType(type), idx));
 }
+#ifndef NDEBUG
+bool TypeSystemSwiftTypeRef::ShouldSkipValidation(opaque_compiler_type_t type) {                                                                               \
+  // NSNotificationName is a typedef to a NSString in clang type, but it's a 
+  // struct in SwiftASTContext. Skip validation in this case.
+  // $s10Foundation12NotificationV4NameaD is a typealias to NSNotificationName,
+  // so we skip validation in that casse as well.
```
Comment (kastiglione) [inline lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp:3478]: It's a shame we don't have a comment saying why `NSDecimal` is here.

Comment (augusto2112) [reply lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp:3478]: Yes, I'm not sure why NSDecimal is skipped, it might be for the same reason but I couldn't find its definition to make sure.
